### PR TITLE
MeZO training pipeline with an example of a small fc layer model on MNIST

### PR DESCRIPTION
**Pull request by sachin-nntrainer**
This PR is an updated code of #3762. It contains the suggested changes. However, I faced a few issues with the code design. 

1. I'm using opt->updateWeightsMeZO(param_ptrs, seed, loss_plus, loss_minus) in 11c5d459dbd60034182c43370795c6dcbcfc401a to updates weights. I feel this does not go well with nntrainer's current logic for updating weights (like backprop). Maybe i should implement something similar to backwarding_op for MeZO and update weights through  RunLayerContext.
2. I faced an error (invalid arg) will setting cross_softmax as loss function. I am yet to debug this. For DEMO purpose i have set MSE as loss function. 

Requesting the reviewer's opinion on this withstanding issues and current changes made to the code. 

Thankyou!

Signed-off-by: Sachin Singh sachin.3@samsung.com

### CODE REVIEW

File: nntrainer/optimizers/MeZO.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  for (const auto *ptr : weights) {
+
+    float *data = (float *)(ptr->getData());
+    size_t param_size = ptr->getDim().getDataLen();
+
+    for (size_t i = 0; i < param_size; ++i) {
+
+      data[i] += lr * normal_dist(gen) * projected_grad;
+    }
+  }
```
Comment (EunjuYang) [inline nntrainer/optimizers/MeZO.cpp:56]: **Bug (Critical):** The update direction is inverted. According to the MeZO paper, the weight update rule is:

```
θ = θ - lr * projected_grad * z
```

But this line does `data[i] += lr * z * projected_grad`, which performs gradient **ascent** (increasing the loss) instead of gradient **descent**.

This should be:
```cpp
data[i] -= lr * normal_dist(gen) * projected_grad;
```

This likely explains why convergence may be poor or non-existent.

File: nntrainer/models/neuralnet.cpp [hunk 2]
```diff
@@ -1370,10 +1375,35 @@ int NeuralNetwork::train_run(
...
+        perturbParameters(-2 * mezo_epsilon, seed);
+        forwarding(true, stop_cb, stop_user_data);
+        float loss_minus = getLoss();
+
+        perturbParameters(mezo_epsilon, seed);
+        float projected_grad = (loss_plus - loss_minus) / (2.0f * mezo_epsilon);
+
+        auto param_ptrs = getParameterPointers();
+        opt->updateWeightsMeZO(param_ptrs, seed, loss_plus, loss_minus);
+
```
Comment (EunjuYang) [inline nntrainer/models/neuralnet.cpp:1401]: `projected_grad` is computed here but never used — `updateWeightsMeZO()` recomputes it internally from `loss_plus` and `loss_minus`. This is dead code and should be removed.

File: nntrainer/models/neuralnet.cpp [hunk 3]
```diff
@@ -1370,10 +1375,35 @@ int NeuralNetwork::train_run(
...
+      } else {
+        ml_logi("train for iteration");
+        forwarding(true, stop_cb, stop_user_data);
+        backwarding(iter++, stop_cb, stop_user_data);
+      }
       // To avoid unconsidered memory leak, we need to clear the cache
```
Comment (EunjuYang) [inline nntrainer/models/neuralnet.cpp:1407]: **Design concern:** Embedding the entire MeZO training loop inside `NeuralNetwork::train_run()` with a type-check branch tightly couples the model to a specific optimizer. Every new gradient-free optimizer (e.g., CMA-ES, evolutionary strategies) would require modifying `NeuralNetwork` again.

**Suggested approach:** Consider adding a `trainStep()` virtual method to the `Optimizer` interface:

```cpp
// In Optimizer base class
virtual bool requiresBackprop() const { return true; }

virtual void trainStep(
  std::function<void()> forward_fn,
  std::function<float()> get_loss_fn,
  std::vector<Tensor*> &params
) { /* default: no-op, backprop optimizers use existing path */ }
```

Then MeZO overrides this to encapsulate all its logic (perturbation, dual forward passes, weight update) inside the optimizer class. `NeuralNetwork::train_run()` simply checks `requiresBackprop()`:

```cpp
if (opt->requiresBackprop()) {
  forwarding(...);
  backwarding(iter++, ...);
} else {
  auto params = getParameterPointers();
  opt->trainStep([&]{ forwarding(...); }, [&]{ return getLoss(); }, params);
  iter++;
}
```

This way:
- `NeuralNetwork` doesn't need to know about MeZO internals
- `perturbParameters()` moves into the MeZO class where it belongs
- `OptimizerWrapped` doesn't need MeZO-specific methods (`getMeZOEpsilon`, `updateWeightsMeZO`)
- Future gradient-free optimizers can follow the same pattern

You mentioned this concern in the PR description — I think the direction of implementing something like `backwarding_op` for MeZO is exactly right.

File: nntrainer/models/neuralnet.cpp [hunk 4]
```diff
@@ -1370,10 +1375,35 @@ int NeuralNetwork::train_run(
...
+        perturbParameters(mezo_epsilon, seed);
+        float projected_grad = (loss_plus - loss_minus) / (2.0f * mezo_epsilon);
+
+        auto param_ptrs = getParameterPointers();
+        opt->updateWeightsMeZO(param_ptrs, seed, loss_plus, loss_minus);
+
+      } else {
+        ml_logi("train for iteration");
+        forwarding(true, stop_cb, stop_user_data);
+        backwarding(iter++, stop_cb, stop_user_data);
```
Comment (EunjuYang) [inline nntrainer/models/neuralnet.cpp:1405]: `iter++` is missing in the MeZO branch. In the backprop path, `iter` is incremented via `backwarding(iter++, ...)`. Without this, iteration-dependent logic (logging, LR scheduling, etc.) won't work correctly.

Add `iter++;` somewhere in this branch.

File: nntrainer/models/neuralnet.cpp [hunk 5]
```diff
@@ -1894,4 +1924,38 @@ void NeuralNetwork::exports(const ml::train::ExportMethods &method,
...
+    nullptr);
+  return params;
+}
+
+void NeuralNetwork::perturbParameters(float epsilon, int seed) {
+  auto param_ptrs = getParameterPointers();
+  std::mt19937 gen(seed);
+  std::normal_distribution<float> normal_dist(0.0f, 1.0f);
+
+  for (const auto *ptr : param_ptrs) {
```
Comment (EunjuYang) [inline nntrainer/models/neuralnet.cpp:1949]: `getData()` returns `const void*`, so this C-style cast `(float *)(ptr->getData())` silently casts away constness. This is technically undefined behavior if the underlying memory is const.

Consider using a non-const accessor if one exists (e.g., `getData<float>()` or a mutable variant), or at the very least use `const_cast` explicitly to make the intent clear. The same issue exists in `MeZO.cpp:52`.

File: nntrainer/models/neuralnet.cpp [hunk 6]
```diff
@@ -1267,7 +1268,11 @@ int NeuralNetwork::train(const std::vector<std::string> &values,
   model_graph.setBatchSize(
     std::get<props::TrainingBatchSize>(model_flex_props));
 
-  status = allocate(ExecutionMode::TRAIN);
+  if (opt->getType() == MeZO::type) {
+    status = allocate(ExecutionMode::INFERENCE);
```
Comment (EunjuYang) [inline nntrainer/models/neuralnet.cpp:1272]: Using `ExecutionMode::INFERENCE` to avoid gradient allocation is a smart optimization, but hardcoding the optimizer type check here is fragile.

Consider letting the optimizer declare its required execution mode:
```cpp
// In Optimizer base class
virtual ExecutionMode getRequiredExecutionMode() const {
  return ExecutionMode::TRAIN;
}
```
Then this becomes simply:
```cpp
status = allocate(opt->getRequiredExecutionMode());
```

File: nntrainer/optimizers/optimizer_wrapped.cpp [hunk 7]
```diff
@@ -140,4 +141,31 @@ nntrainer::LearningRateScheduler *OptimizerWrapped::getLearningRateScheduler() {
...
+    MeZO *mezo_optimizer = static_cast<MeZO *>(optimizer.get());
+    return mezo_optimizer->getEpsilon();
+  } else {
+    // For non-MeZO optimizers, return a default value
+    ml_logw("getEpsilon called on non-MeZO optimizer, returning 0.0f");
+    return 0.0f;
+  }
+}
+
 } // namespace nntrainer
```
Comment (EunjuYang) [inline nntrainer/optimizers/optimizer_wrapped.cpp:171]: Adding MeZO-specific methods (`updateWeightsMeZO`, `getMeZOEpsilon`) to the generic `OptimizerWrapped` class breaks the abstraction. Every new optimizer with custom behavior would require adding new methods here.

If the `trainStep()` approach described in my other comment is adopted, these methods become unnecessary — all MeZO-specific logic stays inside the `MeZO` class, and `OptimizerWrapped` remains a clean, generic wrapper.

File: nntrainer/optimizers/MeZO.h
Comment (EunjuYang) [inline nntrainer/optimizers/MeZO.h:1]: **Naming convention:** Other optimizers follow lowercase naming: `sgd.h`, `adamw.h`, `lion.h`. For consistency, consider renaming to `mezo.h` / `mezo.cpp`.

Similarly, the enum value `ML_TRAIN_OPTIMIZER_TYPE_MeZO` should be `ML_TRAIN_OPTIMIZER_TYPE_MEZO` to follow C enum conventions (all caps).

File: nntrainer/app_context.cpp [hunk 8]
```diff
@@ -31,6 +31,7 @@
 #include <adamw.h>
 #include <lion.h>
 #include <sgd.h>
+#include<MeZO.h>
```
Comment (EunjuYang) [inline nntrainer/app_context.cpp:34]: Nit: Missing space after `#include`.
```cpp
#include <MeZO.h>
```

File: Applications/MeZO/jni/main.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+      model->addLayer(layer);
+    }
+
+    // Set the optimizer
+    auto optimizer = ml::train::createOptimizer(
+      "MeZO", {"MeZO_learning_rate=0.00069", " MeZO_epsilon=0.021"});
+    model->setOptimizer(std::move(optimizer));
+
+    // Set model properties
+    model->setProperty({"epochs=5", "batch_size=32"});
```
Comment (EunjuYang) [inline Applications/MeZO/jni/main.cpp:264]: There is a leading space in `" MeZO_epsilon=0.021"` which may cause a property parsing failure or unexpected behavior.
```cpp
auto optimizer = ml::train::createOptimizer(
  "MeZO", {"MeZO_learning_rate=0.00069", "MeZO_epsilon=0.021"});
```

File: nntrainer/app_context.cpp [hunk 10]
```diff
@@ -108,8 +109,8 @@
 #include <nnstreamer_layer.h>
 #endif
 
-/// add #ifdef across platform
-static std::string solib_suffix = ".so";
+  /// add #ifdef across platform
+  static std::string solib_suffix = ".so";
```
Comment (EunjuYang) [inline nntrainer/app_context.cpp:113]: These indentation changes are unrelated to MeZO. Please revert to keep the diff clean.

**Review by EunjuYang — CHANGES_REQUESTED** [review]
## Review Summary

Thanks for working on MeZO support — zeroth-order optimization is a great addition for on-device training where backpropagation memory is a constraint. The overall MeZO algorithm flow (perturb → forward → estimate gradient → update) is implemented correctly in spirit, but there are a few issues to address before this is ready.

### Critical

1. **Weight update sign is inverted** (`MeZO.cpp:56`): The code does `+=` but MeZO requires `θ = θ - lr * projected_grad * z`, so it should be `-=`. Currently performing gradient ascent.
2. **`projected_grad` computed but unused** in `neuralnet.cpp` — it's recomputed inside `updateWeightsMeZO()`.
3. **`iter++` missing** in the MeZO training branch.

### Architecture

The main design concern is that MeZO-specific logic (perturbation, dual forward passes, custom weight update) is spread across `NeuralNetwork`, `OptimizerWrapped`, and `MeZO` classes. I've suggested a `trainStep()` virtual method pattern on the `Optimizer` interface that would:
- Keep all MeZO logic inside the `MeZO` class
- Remove MeZO-specific methods from `OptimizerWrapped`
- Remove `perturbParameters()` / `getParameterPointers()` from `NeuralNetwork`
- Make it easy to add future gradient-free optimizers

See the inline comment on `neuralnet.cpp:1378-1407` for the detailed proposal.

### Other items
- **const violation**: C-style cast discards `const` from `getData()` — potential UB
- **Naming convention**: `MeZO.h/cpp` should follow project convention (`mezo.h/cpp`)
- **Leading space** in optimizer property string `" MeZO_epsilon=0.021"`
- **Unrelated whitespace changes** in `app_context.cpp`
- **No unit tests** for the new optimizer
- **Missing newline** at end of `main.cpp` and `meson.build`

### Positive notes
- The MNIST example is clean and well-structured
- Using `ExecutionMode::INFERENCE` to skip gradient allocation is a good insight
- `applyGradient()` returning empty and `getOptimizerVariableDim()` returning `{}` correctly reflects that MeZO doesn't use per-parameter optimizer state

Looking forward to the next iteration!

**Comment by sachin-nntrainer** [comment]
Thankyou @EunjuYang for a Comprehensive review. I will work towards fixing them !
